
---
title: 'How much PCR amplification is biasing DNA metabarcoding results ? - data processing'
author:
  - name: Sylvain Moinard
    institute: leca
    email: sylvain.moinard@univ-grenoble-alpes.fr
    correspondence: true
  - name: Didier Piau
    institute: ijf
  - name: Frédéric Laporte
    institute: leca
  - name: Delphine Rioux
    institute: leca
  - name: Christelle Gonindard-Melodelima
    institute: leca
  - name: Eric Coissac
    institute: leca
    email: eric.coissac@metabarcoding.org 
    correspondence: true
institute:
  - leca: Univ. Grenoble-Alpes, Univ. Savoie Mont Blanc, CNRS, LECA, 38000, Grenoble, France.
  - ijf: Univ. Grenoble-Alpes, CNRS, Institut Joseph Fourier, 38000, Grenoble, France.
output: html_document
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(dplyr)
library(readxl)
library(tidyverse)

#install_url("https://git.metabarcoding.org/obitools/obitools4/robireadfasta")
library(ROBIFastread)
```

# Data related to the manuscript

## Metabarcodes for marker Sper01 (gh)

Salvia pratensis has a different barcode as in the database (one insertion)

```{r open reference}
barcodes1 <- read_excel("data/new_plant_positive_control.xlsx")[,c(2,4:6)]
names(barcodes1) <- c("species_name", "sequence", "length", "gc_content")

barcodes1[barcodes1$species_name ==
            "Salvia_pratensis",]$sequence <- "atcctgttttctcaaaacaaaggttcaaaaaacgaaaaaaaaaaag"

```

## DNA extraction

```{r extraction}
extraction <- read_excel(
  "data/dosage extraction plantes.xlsx")
extraction <- extraction[-nrow(extraction),]
extraction <- extraction[,-(4:9)]
```


## ddPCR data

Open, clean, summarise ddPCR data

```{r ddPCR data}
#First file
dataddpcr1 <- read.csv("data/QMETABAR_MOINARD1204_manual.csv",
                       header=FALSE)
colnames(dataddpcr1) <- dataddpcr1[1,]
dataddpcr1 <- dataddpcr1[-1,c(1,7,8,16,22)]

dataddpcr1$Sample <- NA
dataddpcr1$sp <- NA
dataddpcr1$Dilu <- NA

dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "01",]$Sample <- 100
dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "01",]$sp <- "Spk1"

dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "02",]$Sample <- 14
dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "02",]$sp <- "Pab"

dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "03",]$Sample <- 16
dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "03",]$sp <- "Lxy"

dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "04",]$Sample <- 18
dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "04",]$sp <- "Aal"

dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "05",]$Sample <- 20
dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "05",]$sp <- "Aca"

dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "06",]$Sample <- 22
dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "06",]$sp <- "Bme"

dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "07",]$Sample <- -1
dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "07",]$sp <- "Neg"

dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "08",]$Sample <- 24
dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "08",]$sp <- "Rca"

dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "09",]$Sample <- 26
dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "09",]$sp <- "Cbp"

dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "10",]$Sample <- 28
dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "10",]$sp <- "Gro"

dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "11",]$Sample <- 30
dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "11",]$sp <- "Rfe"

dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "12",]$Sample <- 32
dataddpcr1[substr(dataddpcr1$Well, 2, 3) == "12",]$sp <- "Lco"

#Corrections

dataddpcr1[dataddpcr1$Well %in% c("H07"),]$Sample <- 100
dataddpcr1[dataddpcr1$Well %in% c("H07"),]$sp <- "Spk1"

dataddpcr1[dataddpcr1$Well %in% c("H06"),]$Sample <- 20
dataddpcr1[dataddpcr1$Well %in% c("H06"),]$sp <- "Aca"

dataddpcr1[dataddpcr1$Well %in% c("H01", "A12", "B12"),]$Sample <- -1
dataddpcr1[dataddpcr1$Well %in% c("H01", "A12", "B12"),]$sp <- "Neg"

dataddpcr1[dataddpcr1$Well %in% c("H05"),]$Sample <- 22
dataddpcr1[dataddpcr1$Well %in% c("H05"),]$sp <- "Bme"

dataddpcr1[dataddpcr1$Well %in% c("A07", "B07"),]$Sample <- 32
dataddpcr1[dataddpcr1$Well %in% c("A07", "B07"),]$sp <- "Lco"

dataddpcr1[substr(dataddpcr1$Well, 1, 1) %in% c("A", "B"),]$Dilu <- 1
dataddpcr1[substr(dataddpcr1$Well, 1, 1) %in% c("C", "D"),]$Dilu <- 2
dataddpcr1[substr(dataddpcr1$Well, 1, 1) %in% c("E", "F"),]$Dilu <- 3
dataddpcr1[substr(dataddpcr1$Well, 1, 1) %in% c("G", "H"),]$Dilu <- 4

dataddpcr1$Concentration <- as.numeric(dataddpcr1$Concentration)
dataddpcr1$AcceptedDroplets <- as.numeric(dataddpcr1$AcceptedDroplets)
dataddpcr1$Negatives <- as.numeric(dataddpcr1$Negatives)

#Manual calls :

#Volume of a droplet
#using the formula to determine concentration
Vdroplet <- log(dataddpcr1[dataddpcr1$Status == "OK",]$AcceptedDroplets/
        dataddpcr1[dataddpcr1$Status == "OK",]$Negatives)/
  dataddpcr1[dataddpcr1$Status == "OK",]$Concentration #ul

vdroplet <- mean(Vdroplet)

dataddpcr1[dataddpcr1$Status == "CHECK",]$Concentration <-
  log(dataddpcr1[dataddpcr1$Status == "CHECK",]$AcceptedDroplets/
        dataddpcr1[dataddpcr1$Status == "CHECK",]$Negatives)/vdroplet

#Second file

dataddpcr2 <- read.csv("data/QMETABAR_MOINARD_140422_manual.csv",
                        header=FALSE)
colnames(dataddpcr2) <- dataddpcr2[1,]
dataddpcr2 <- dataddpcr2[-1,c(1,4,6,7,8,16,22)]

colnames(dataddpcr2)[3] <- "Dilu" #Target
dataddpcr2$sp <- NA
dataddpcr2$Sample <- as.numeric(dataddpcr2$Sample)

dataddpcr2[dataddpcr2$Sample == 100,]$sp <- "Spk1"
dataddpcr2[dataddpcr2$Sample == -1,]$sp <- "Neg"
dataddpcr2[dataddpcr2$Sample == 4,]$sp <- "Spr"
dataddpcr2[dataddpcr2$Sample == 6,]$sp <- "Ptr"
dataddpcr2[dataddpcr2$Sample == 8,]$sp <- "Rac"
dataddpcr2[dataddpcr2$Sample == 10,]$sp <- "Cbe"
dataddpcr2[dataddpcr2$Sample == 12,]$sp <- "Fex"
dataddpcr2[dataddpcr2$Sample == 20,]$sp <- "Aca"
dataddpcr2[dataddpcr2$Sample == 32,]$sp <- "Lco"

dataddpcr2$Concentration <- as.numeric(dataddpcr2$Concentration)
dataddpcr2$AcceptedDroplets <- as.numeric(dataddpcr2$AcceptedDroplets)
dataddpcr2$Negatives <- as.numeric(dataddpcr2$Negatives)

#Manual calls:
Vdroplet2 <- log(dataddpcr2[dataddpcr2$Status == "OK",]$AcceptedDroplets/
                      dataddpcr2[dataddpcr2$Status == "OK",]$Negatives)/
  dataddpcr2[dataddpcr2$Status == "OK",]$Concentration #ul

vdroplet2 <- mean(Vdroplet2)

dataddpcr2[dataddpcr2$Status == "CHECK",]$Concentration <-
  log(dataddpcr2[dataddpcr2$Status == "CHECK",]$AcceptedDroplets/
        dataddpcr2[dataddpcr2$Status == "CHECK",]$Negatives)/vdroplet2

#Merge 2 plates
dataddpcr1$plate <- 1
dataddpcr2$plate <- 2

dataddpcr <- rbind(dataddpcr1, dataddpcr2)

dataddpcr$Dilu <- as.numeric(dataddpcr$Dilu)

#Removing concentrations outliers
dataddpcr$Used <- F

sample_OK <- c(4, 6, 10, 12, 16, 18, 20, 22, 24, 26, 28, 30, 32, 100)

dataddpcr[dataddpcr$Sample == 4,]$Used  <- T
dataddpcr[dataddpcr$Sample == 6 & dataddpcr$Status == "OK",]$Used <- T
dataddpcr[dataddpcr$Sample == 10,]$Used <- T
dataddpcr[dataddpcr$Sample == 12,]$Used <- T
dataddpcr[dataddpcr$Sample == 16,]$Used <- T
dataddpcr[dataddpcr$Sample == 18 &
             dataddpcr$Dilu < 4,]$Used <- T
dataddpcr[dataddpcr$Sample == 20 &
             dataddpcr$Status != "CHECK",]$Used <- T
dataddpcr[dataddpcr$Sample == 22 &
             dataddpcr$Dilu < 4,]$Used <- T
dataddpcr[dataddpcr$Sample == 24 &
             dataddpcr$Dilu < 4 & dataddpcr$Status != "CHECK",]$Used <- T
dataddpcr[dataddpcr$Sample == 26 &
             dataddpcr$Dilu < 4,]$Used <- T
dataddpcr[dataddpcr$Sample == 28 &
             dataddpcr$Dilu < 4,]$Used <- T
dataddpcr[dataddpcr$Sample == 30 &
             dataddpcr$Dilu < 4,]$Used <- T
dataddpcr[dataddpcr$Sample == 32 &
             dataddpcr$Dilu < 4 &
             !(dataddpcr$Well %in%c("E12", "F12")),]$Used <- T
dataddpcr[dataddpcr$Sample == 100 &
             dataddpcr$Dilu < 4 &
             dataddpcr$Status != "CHECK" &
             !(dataddpcr$Well %in%c("H03")),]$Used <- T

#Volumes
Vmix_dd <- 20 #ul
Vdna_dd <- 5 #ul

CtotDNA_ref <- 2.5e-3 *Vdna_dd/Vmix_dd

maxDNA <- 0.25 *Vdna_dd/Vmix_dd #ng/ul
dataddpcr$TotalDNA <- maxDNA/10^(dataddpcr$Dilu-1)

dataddpcr <- dataddpcr %>%
  mutate(Concentration_std = Concentration*CtotDNA_ref/TotalDNA)

summary_dd <- dataddpcr %>% filter(Used & Sample < 100) %>%
  group_by(Sample) %>%
  summarise(sp = sp[1],
            Mean = mean(Concentration*CtotDNA_ref/TotalDNA),
            min = min(Concentration*CtotDNA_ref/TotalDNA),
            max = max(Concentration*CtotDNA_ref/TotalDNA)) %>%
  left_join(extraction %>% dplyr::select(Espece, Sample, `concentration (ng/µl`)) %>%
  rename(CDNA_sample = `concentration (ng/µl`,
         Cmolecules_ref = Mean,
         species_name = Espece) %>%
  dplyr::select(Sample, species_name, sp,
                Cmolecules_ref, CDNA_sample)

# write_csv(dataddpcr, "data/dataddpcr.csv")
```
## Metabarcoding data processing

Processing the data with OBITools version 4
(https://git.metabarcoding.org/obitools/obitools4/obitools4)

Three communities:
* L1 = uniform in Total DNA (renamed T in the paper)
* G1 = Geometric
* U1 = Uniform

### Bash commands

#### Gobipairing

Paired-end data

% gobipairing -F 220930_NB501850_A_L1-4_GWM-1746_R1.fastq.gz -R 220930_NB501850_A_L1-4_GWM-1746_R2.fastq.gz > GWM-1746.paired

% gobipairing -F 220930_NB501850_A_L1-4_GWM-1748_R1.fastq.gz -R 220930_NB501850_A_L1-4_GWM-1748_R2.fastq.gz > GWM-1748.paired

% gobipairing -F 220930_NB501850_A_L1-4_GWM-1750_R1.fastq.gz -R 220930_NB501850_A_L1-4_GWM-1750_R2.fastq.gz > GWM-1750.paired

#### Convert json (made by metabase api) to ngsfilter

% python3 format_ngsfilter_sylvain.py ngs_filter_tube_L1_EC_QMETABAR_210722.json GWM-1746_L1

% python3 format_ngsfilter_sylvain.py ngs_filter_tube_G1_EC_QMETABAR_210722.json GWM-1748_G1

% python3 format_ngsfilter.py ngs_filter_tube_U1_EC_QMETABAR_210722.json GWM-1750_U1

#### Gobimultiplex

Assigns sequence records to the corresponding experiment/sample based on DNA tags and primers

% gobimultiplex -t GWM-1746_L1.ngsfilter -u unassigned46 GWM-1746.paired > GWM_1746.demultiplex

% gobimultiplex -t GWM-1748_G1.ngsfilter -u unassigned48 GWM-1748.paired > GWM_1748.demultiplex

% gobimultiplex -t GWM-1750_U1.ngsfilter -u unassigned50 GWM-1750.paired > GWM_1750.demultiplex

#### Gobiuniq

Groups and dereplicates sequences
Merge the libraries

% gobiuniq -m direction -m sample --on-disk GWM_1746.demultiplex GWM_1748.demultiplex GWM_1750.demultiplex > GWM_Sper01.uniq

#### Gobigrep

Filters sequence file

% gobigrep -c 2 -l 25 -L 75 GWM_Sper01.uniq > GWM_Sper01.grep

#### Gobiclean

Tags a set of sequences for PCR/sequencing errors identification

% gobiclean --save-ratio graphSper01.csv --min-eval-rate 100 --save-graph graph1 GWM_Sper01.grep > GWM_Sper01.clean

#### Gobicount

Reads / Variants

% gobicount GWM-1746.paired
3352859 / 3352859

% gobicount GWM_1746.demultiplex
3043811 / 3043811

% gobicount unassigned46
309048 / 309048

% gobicount GWM-1748.paired
4025048 / 4025048

% gobicount GWM_1748.demultiplex
3683622 / 3683622

% gobicount unassigned48
341426 / 341426


% gobicount GWM-1750.paired
3910259 / 3910259

% gobicount GWM_1750.demultiplex
3618993 / 3618993

% gobicount unassigned50
291266 / 291266


% gobicount GWM_Sper01.uniq
151109 / 10346426

% gobicount GWM_Sper01.grep
28371 / 10202369

% gobicount GWM_Sper01.clean
28371 / 10202369

#### Robireadfasta

Fasta files are read with the R library ROBIFastread available here:
https://git.metabarcoding.org/obitools/obitools4/robireadfasta

### Import Metabarcoding data

The data come from a larger experiment. We remove every replicate containing Spike molecules.

```{r open metabarcoding data}
data_Sper01 <- read_obifasta("data/GWM_Sper01.clean")

weights <- extract_readcount(data_Sper01,
                    key = "obiclean_weight")

reads <- extract_readcount(data_Sper01)
df_reads <- Matrix::summary(reads)
df_reads <- data.frame(Sample = rownames(reads)[df_reads$i],
           id = colnames(reads)[df_reads$j],
           reads = df_reads$x)

df_Sper01 <- Matrix::summary(weights)

df_Sper01 <- data.frame(Sample = rownames(weights)[df_Sper01$i],
           id = colnames(weights)[df_Sper01$j],
           obiclean_weight = df_Sper01$x) %>%
  left_join(df_reads) %>%
  left_join(data_Sper01 %>% dplyr::select(id, sequence)) %>%
  mutate(Community = substr(Sample, 1, 1),
         NegControl = Community == "P")

df_Sper01[df_Sper01$Community == "P", "Community"] <- substr(df_Sper01[df_Sper01$Community == "P",]$Sample, 8, 8)
df_Sper01[df_Sper01$Community == "L", "Community"] <- "T"

df_Sper01$Spike_level <- NA
df_Sper01[!df_Sper01$NegControl,"Spike_level"] <- as.numeric(substr(df_Sper01[!df_Sper01$NegControl,]$Sample, 4, 4))

df_Sper01[df_Sper01$NegControl,"Spike_level"] <- as.numeric(substr(df_Sper01[df_Sper01$NegControl,]$Sample, 12, 12))

df_Sper01 <- df_Sper01 %>% filter(Spike_level == 1) #***

summary_sample <- df_Sper01 %>%
  group_by(Sample) %>%
  summarise(total_count_sample = sum(reads),
            NegControl = NegControl[1],
            Spike_level = Spike_level[1],
            Community = Community[1]) %>%
  arrange(total_count_sample)

summary_species <- df_Sper01 %>%
  group_by(sequence) %>%
  summarise(count = sum(reads)) %>%
  arrange(desc(count))

summary_species$seq_rank <- 1:nrow(summary_species)

df_Sper01 <- df_Sper01 %>% left_join(summary_sample[,1:2]) %>%
  left_join(summary_species) %>%
  left_join(barcodes1) %>%
  arrange(desc(count))

df_Sper01[is.na(df_Sper01$species_name),"species_name"] <- as.character(df_Sper01[is.na(df_Sper01$species_name),]$seq_rank)

# write_csv(df_Sper01, "data/df_Sper01.csv")

```

Export data to Julia for Taqman analysis

```{r export julia}
reads_threshold <- 5000

##Commu U
auxU <- df_Sper01 %>% filter(Community == "U" &
                                   species_name %in% barcodes1$species_name &
                        species_name != "Rumex_acetosa" &
                          total_count_sample > reads_threshold &
                          !NegControl) %>%
  filter(species_name %in%
           c("Fraxinus_excelsior", "Carpinus_betulus", "Capsella_bursa-pastoris")) %>%
  dplyr::select(species_name, obiclean_weight) %>%
  arrange(species_name) %>%
  pivot_wider(names_from = species_name, values_from = obiclean_weight)

matU <- as.matrix(auxU[,-1])
# write.csv(matU, "data/dfSper01U_taq.csv")

##Commu T
auxT <- df_Sper01 %>% filter(Community == "T" &
                                   species_name %in% barcodes1$species_name &
                        species_name != "Rumex_acetosa" &
                          total_count_sample > reads_threshold &
                          !NegControl) %>%
  filter(species_name %in%
           c("Fraxinus_excelsior", "Carpinus_betulus", "Capsella_bursa-pastoris")) %>%
  dplyr::select(species_name, obiclean_weight) %>%
  arrange(species_name) %>%
  pivot_wider(names_from = species_name, values_from = obiclean_weight)
matT <- as.matrix(auxT[,-1])
# write.csv(matT, "data/dfSper01T_taq.csv")

##Commu G
auxG <- df_Sper01 %>% filter(Community == "G" &
                                   species_name %in% barcodes1$species_name &
                        species_name != "Rumex_acetosa" &
                          total_count_sample > reads_threshold &
                          !NegControl) %>%
  filter(species_name %in%
           c("Fraxinus_excelsior", "Carpinus_betulus", "Capsella_bursa-pastoris")) %>%
  dplyr::select(species_name, obiclean_weight) %>%
  arrange(species_name) %>%
  pivot_wider(names_from = species_name, values_from = obiclean_weight)
matG <- as.matrix(auxG[,-1])
# write.csv(matG, "data/dfSper01G_taq.csv")

```

## Taqman qPCR

```{r pipeline taqman}
taqman <- summary_dd %>%
  filter(Sample %in% c(10,12, 26)) %>%
  dplyr::select(Sample, species_name, sp, Cmolecules_ref, CDNA_sample) %>%
  distinct(Sample, .keep_all = TRUE)

#Reference concentration = sample Cbe 10

#3rd level concentration
cref_taqman <- 2.484696e-03 #ng/ul total DNA

#5-fold serial dilution
#Concentration (total DNA) of each dilution level
cref_taqman*c(25,5,1,1/5,1/25)

#Copies/ul at cref_taqman in the sample
taqman[taqman$Sample == 10,]$Cmolecules_ref*cref_taqman/CtotDNA_ref

Vdna_taq <- 5 #ul
Vmix_taq <- 25 #ul

#Copies/ul at cref_taqman in the taqman mix
taqman[taqman$Sample == 10,]$Cmolecules_ref*
  cref_taqman/CtotDNA_ref*Vdna_taq/Vmix_taq

moc_well <- taqman[taqman$Sample == 10,]$Cmolecules_ref*
  cref_taqman*25/CtotDNA_ref*Vdna_taq

moc_tot <- moc_well*35 #Molecules for the whole experiment of each species

taqman1 <- taqman
taqman1$mix <- 1
taqman1$copies_ul <- moc_tot/175*Vdna_taq/Vmix_taq
taqman1$CDNA_well <- taqman1$copies_ul*
  CtotDNA_ref/taqman$Cmolecules_ref

taqman2 <- taqman
taqman2$mix <- 2
taqman2$copies_ul <- taqman1$copies_ul/5
taqman2$CDNA_well <- taqman1$CDNA_well/5

taqman3 <- taqman
taqman3$mix <- 3
taqman3$copies_ul <- taqman2$copies_ul/5
taqman3$CDNA_well <- taqman2$CDNA_well/5

taqman4 <- taqman
taqman4$mix <- 4
taqman4$copies_ul <- taqman3$copies_ul/5
taqman4$CDNA_well <- taqman3$CDNA_well/5

taqman5 <- taqman
taqman5$mix <- 5
taqman5$copies_ul <- taqman4$copies_ul/5
taqman5$CDNA_well <- taqman4$CDNA_well/5

taqman <- rbind(taqman1, taqman2, taqman3, taqman4, taqman5)

# write_csv(taqman, "data/taqman.csv")
```

## Tags for metabarcoding PCR

```{r metabarcoding tags}
primers1 <- read.table("data/DEL0455463409_bis.csv",
                       header = T, sep = ",", na.strings = c("N/A", "-"))
primers1 <- primers1[,-which(is.na(primers1[1,]))]

primers2 <- read.table("data/DEL0455483280_bis.csv",
                       header = T, sep = ",", na.strings = c("N/A", "-"))
primers2 <- primers2[,-which(is.na(primers2[1,]))]

primers <- rbind(primers1, primers2)

primers$Well.col <- as.numeric(substr(primers$Well.position, 2, 3))
primers$Well.row <- substr(primers$Well.position, 1, 1)
primers$Marker <- substr(primers$Name, 1, 6)
primers$Direction <- substr(primers$Name, 7, 7)

primers$Sper_seq <- NULL
primers[primers$Marker == "SPER01" & primers$Direction == "F",
        "Sper_seq"] <- "GGGCAATCCTGAGCCAA"
primers[primers$Marker == "SPER01" & primers$Direction == "R",
        "Sper_seq"] <- "CCATTGAGTCTCTGCACCTATC"

colnames(primers)[8] <- "Sequence"

primers$Tag <- sapply(1:nrow(primers),
                      function(i){
                        substr(primers[i,]$Sequence,1,
                               str_locate(
                                 primers[i,]$Sequence,
                                 primers[i,]$Sper_seq)-1)})

primers$Tag8 <- substr(primers$Tag, nchar(primers$Tag)-8+1,
                       nchar(primers$Tag))
```


