---
title: 'Towards quantitative DNA Metabarcoding: A method to overcome PCR amplification bias'
author:
  - name: Sylvain Moinard
    institute: leca
    email: sylvain.moinard@univ-grenoble-alpes.fr
  - name: Didier Piau
    institute: ijf
  - name: Frédéric Laporte
    institute: leca
  - name: Delphine Rioux
    institute: leca
  - name: Pierre Taberlet
    institute: leca
  - name: Christelle Gonindard-Melodelima
    institute: leca
  - name: Eric Coissac
    institute: leca
    email: eric.coissac@metabarcoding.org 
    correspondence: true
institute:
  - leca: Univ. Grenoble-Alpes, Univ. Savoie Mont Blanc, CNRS, LECA, 38000, Grenoble, France.
  - ijf: Univ. Grenoble-Alpes, CNRS, Institut Joseph Fourier, 38000, Grenoble, France.
output:
  html_document: 
    df_print: paged
    number_sections: yes
    pandoc_args:
      - '--lua-filter=style/scholarly-metadata.lua'
      - '--lua-filter=style/author-info-blocks.lua'
    css: style/splendor.css
---


This project contains the data and analysis scripts linked to the paper:

***Towards quantitative DNA Metabarcoding: A method to overcome PCR amplification bias***

05/10/23: The paper is available as a preprint. Cite as:

*Towards quantitative DNA Metabarcoding: A method to overcome PCR amplification bias
Sylvain Moinard, Didier Piau, Frederic Laporte, Delphine Rioux, Pierre Taberlet, Eric Coissac, Christelle Gonindard-Melodelima
bioRxiv 2023.10.03.560640; doi: https://doi.org/10.1101/2023.10.03.560640*

## Directory contents

### Data

#### data/ddpcr

ddPCR results generated by Quantasoft (BioRad) software:

- `QMETABAR_MOINARD_120422_manual.csv` : Plate 1.
- `QMETABAR_MOINARD_140422_manual.csv` : Plate 2.

#### data/extraction

- `dosage_extraction_plantes.xlsx` : Information about plants extraction and DNA concentration.

#### data/metabarcoding

Some files are too large to be stored in this git but the pipeline allows to reproduce all these files with the OBITools softwares. The pipeline is shown in file data_processing.Rmd, section Metabarcoding data processing

Raw paired-end sequencing data produced during metabarcoding experiment:

- `220930_NB501850_A_L1-4_GWM-1746_R1.fastq.gz` : community M_T, reads 1
- `220930_NB501850_A_L1-4_GWM-1746_R2.fastq.gz` : community M_T, reads 2
- `220930_NB501850_A_L1-4_GWM-1748_R1.fastq.gz` : community M_G, reads 1
- `220930_NB501850_A_L1-4_GWM-1748_R2.fastq.gz` : community M_G, reads 2
- `220930_NB501850_A_L1-4_GWM-1750_R1.fastq.gz` : community M_U, reads 1
- `220930_NB501850_A_L1-4_GWM-1750_R2.fastq.gz` : community M_U, reads 2

Information about tags added during metabarcoding experiment:

- `DEL0455463409_bis.csv` : Forward tags
- `DEL0455483280_bis.csv` : Reverse tags

Data processed with the obitools softwares. All the bash scripts are available in file data_processing.Rmd.

First, generate .paired files from raw reads data with gobipairing (not stored on git).

Then, run script format_ngsfilter.py in repo data/metabarcoding using .json files:

.json contain all the information about the PCR plates.

- `ngs_filter_tube_G1_EC_QMETABAR_210722.json` : community M_G
- `ngs_filter_tube_T1_EC_QMETABAR_210722.json` : community M_T
- `ngs_filter_tube_U1_EC_QMETABAR_210722.json` : community M_U

produced files:

- `GWM-1746_T1.ngsfilter` : community M_T
- `GWM-1748_G1.ngsfilter` : community M_G
- `GWM-1750_U1.ngsfilter` : community M_U

Generate .demultiplex files using gobidemultiplex (not stored on git).

Generate .uniq file containing data of the three communities using gobuniq:

- `GWM_Sper01.uniq`

Generate .grep file with filtered sequences:

- `GWM_Sper01.grep`

Generate .clean file to filter erroneous sequences:

- `GWM_Sper01.clean`

#### data/plants_info

- `new_plant_positive_control.xlsx` : information about plant sampling and barcodes for the Sper01 marker.

#### data/qpcr_sybr

- `qpcr_Quantification_Amplification_Results_SYBR.csv` : qPCR RFU data used to draw Figure 1, generated by CFX Maestro Software (BioRad) from SybrGreen qPCR data.

#### data/qpcr_taqman

Data produced by CFX Maestro Software (BioRad) from Taqman qPCR data.

- `taqman_SPER01_kinetics_res.csv` : RFU data.
- `taqman_SPER01_Cq_res.csv` : quantification data.

#### data/export_to_julia

This repository contains data formatted to be used in Julia.

Reads numbers:

- `reads_G.csv` : community M_G
- `reads_T.csv` : community M_T
- `reads_U.csv` : community M_U

Initial quantity of target molecules:

- `qty_initG.csv` : community M_G
- `qty_initT.csv` : community M_T
- `qty_initU.csv` : community M_U

Simulated communities to produce Supplementary Figure 3. Each file contains the composition of 100 communities of 13 species, with different initial quantities of molecules M0tot.

- `sim100_m0tot_1.csv` : M0tot = 2.5e4
- `sim100_m0tot_2.csv` : M0tot = 2.5e5, as in community M_U
- `sim100_m0tot_3.csv` : M0tot = 2.5e6

#### data/export_to_r

This repository contains results produced in Julia that are then exported to R for further analysis.

efficiencies_U files contain amplification efficiencies inferred from M_U for different values of carrying capacity K. Species order is specified in metabar_bias.Rmd.

- `efficiencies_U_K11.csv` : K = 1.2e11
- `efficiencies_U_K12.csv` : K = 1.2e12
- `efficiencies_U_K13.csv` : K = 1.2e13, which is the estimated value computed from reaction conditions
- `efficiencies_U_K14.csv` : K = 1.2e14

prop_inferG files contain proportions in M_G inferred with corresponding efficiencies (see above) for different values of carrying capacity K. Species order is specified in metabar_bias.Rmd.

- `prop_inferG_K11.csv` : K = 1.2e11
- `prop_inferG_K12.csv` : K = 1.2e12
- `prop_inferG_K13.csv` : K = 1.2e13, which is the estimated value computed from reaction conditions
- `prop_inferG_K14.csv` : K = 1.2e14

prop_inferT files contain proportions in M_T inferred with corresponding efficiencies (see above) for different values of carrying capacity K. Species order is specified in metabar_bias.Rmd.

- `prop_inferT_K11.csv` : K = 1.2e11
- `prop_inferT_K12.csv` : K = 1.2e12
- `prop_inferT_K13.csv` : K = 1.2e13, which is the estimated value computed from reaction conditions
- `prop_inferT_K14.csv` : K = 1.2e14


Inference results (proportions of each species) for 100 simulated communities used in Supplementary Figure 3. Different initial quantities of molecules M0tot.

- `res_commu1.csv` : M0tot = 2.5e4
- `res_commu2.csv` : M0tot = 2.5e5, as in community M_U
- `res_commu3.csv` : M0tot = 2.5e6

#### data/processed_data

These files are generated in data_processing.Rmd to produce ready-to-use data in the main analysis file, metabar_bias.Rmd.

- `dataddpcr.csv` : summarising ddPCR data.
- `df_Sper01.csv` : processed metabarcoding data for the 13 species of interest
- `extraction.csv` : summarising DNA extraction data.
- `quantification_taqman.csv` : Taqman qPCR quantification results.
- `taqman_kinetics.csv` : Taqman qPCR RFU kinetics.
- `preparing_taqman.csv` : information about the design of Taqman experiment.


### Data analysis

#### data_processing

- `data_processing.Rmd` : This file is used to prepare the data for the analysis performed in metabar_bias.Rmd.

Corresponding HTML file: [`data_processing.html`](data_processing.html)

#### metabar_bias

- `metabar_bias.Rmd` : This is the main analysis file.

Corresponding HTML file: [`metabar_bias.html`](metabar_bias.html)

#### Taqman_design

- `Taqman_design.Rmd` : This file is used to design Taqman qPCR probes for the three species chosen for this analysis.

Corresponding HTML file: [`Taqman_design.html`](Taqman_design.html)

#### scripts_Julia

This repository contains the Julia files used to infer PCR efficiencies and proportions of the different species. Their use is specified in metabar_bias.Rmd.

- `metabar_bias_functions.jl` : Simulation and optimization functions used in the inference files below.
- `inference_efficiencies_ps.jl` : Inference of efficiencies (Lambda_s) from M_U data and then of proportions in M_T and M_G. Allows to test inference with different values of carrying capacity K.
- `inference_simulated_communities.jl` : Inference of proportions with simulated communities to produce Supplementary Figure 3.
